mlir/lib/Dialect/Presburger/Parser.cpp

```diff
@@ -63,6 +63,10 @@ std::string Token::name(Token::Kind kind) {
     return "\"and\"";
   case Token::Kind::Or:
     return "\"or\"";
+  case Token::Kind::Arrow:
```

Comment:
I think that this method can currently read too far, i.e. out of the space of the provided `buffer`. Should this be forbidden or is this a problem for the owner of the `Lexer`?

Reply:
There is no check if `curPtr` points further than the `StringRef`. Now that I think of it, the custom parser will only be called on the string inside `<"` `">` and therefore will always throw an error if it reaches `"` as this is an unexpected character.

Reply:
Oh, turns out that I commented on the wrong method, sorry. It should have been in `Lexer::nextToken`, sorry.

Reply:
I feel it would be better to handle this case explicitly.

Reply:
I'm not sure if this can be done without rewriting the `Lexer`. `Lexer::nextToken` returns a `Token` and changing this to `LogicalResult` might be cumbersome. Everything would need to be changed for a case that cannot happen when the `PresburgerParser` is correctly used (it is only supposed to be used as a custom Attribute parser). The "correct usage" should, of course, be documented somewhere.  

Note:
Explicitly enforcing that `curPtr` is only dereferenced within the bounds of `buffer` would require checks all over the place, i.e. either on every increment or on every dereferencing.

Reply:
Sorry, I should've written more clearly. IMO, an assert failure here in case of out of bounds is sufficient, no need to return a `LogicalResult`.

I see your point about checks all over the place. However, I noticed that `ArrayRef`, and hence `StringRef`, assert this in the `operator[]`. Therefore, switching to indices instead of pointers should resolve this transparently.

Reply:
No worries. 
Nice to know that they use assertions. I will change it then.

Reply:
`Lexer::next` always calls `Lexer::nextToken`. To make sure we do not run into an out of bounds assertion, I'll add a check and return an EOF token.